code.js

```diff
@@ -1,3 +1,59 @@
 function tsp_hk(distance_matrix) {
-    return -1;
+    if (distance_matrix.length <= 1) {
+        return 0;
+    }
+    
+    // Resets cache every time the function is called
+    let cache = {};
+
+    let cities = Array.from({ length: distance_matrix.length }, (_, i) => i);
+
+    let minTourLength = Infinity;
+
+    // Loop through the matrix to find the shortest path
+    for (let start = 0; start < cities.length; start++) {
+        let tourLength = heldKarp(distance_matrix, start, cities, cache);
+        minTourLength = Math.min(minTourLength, tourLength);
+    }
+
+    return minTourLength === Infinity ? 0 : minTourLength;
 }
+
+function heldKarp(dm, start, cities, cache) {
+    let key = JSON.stringify([cities, start]);
```

Comment:
What if `cities` is in a different order? Shouldn't that get you the same cache entry?